File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/MainFragment.java
package com.remotearthsolutions.expensetracker.fragments;

import android.animation.ValueAnimator;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.databinding.FragmentMainBinding;
import com.remotearthsolutions.expensetracker.fragments.home.DateFilterButtonClickListener;
import com.remotearthsolutions.expensetracker.fragments.home.HomeFragment;
import com.remotearthsolutions.expensetracker.utils.Constants;
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils;


public class MainFragment extends Fragment implements DateFilterButtonClickListener.Callback {

    private FragmentMainBinding binding;
    private MainFragmentPagerAdapter pagerAdapter;
    private ActionBar actionBar;
    private String[] tabTitles = new String[]{getString(R.string.home), getString(R.string.transactions), getString(R.string.accounts), getString(R.string.dashboard),};

    private static HomeFragment homeFragment;
    private static AllExpenseFragment allExpenseFragment;
    private static DashboardFragment dashboardFragment;
    private static AccountsFragment accountsFragment;
    private int dateContainerHeight = -1;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false);

        pagerAdapter = new MainFragmentPagerAdapter(getChildFragmentManager());
        binding.viewpager.setOffscreenPageLimit(tabTitles.length - 1);
        binding.viewpager.setAdapter(pagerAdapter);
        binding.viewpager.addOnPageChangeListener(viewPagerPageChangeListener);
        binding.navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);

        DateFilterButtonClickListener dateFilterButtonClickListener = new DateFilterButtonClickListener(this);
        binding.nextDateBtn.setOnClickListener(dateFilterButtonClickListener);
        binding.previousDateBtn.setOnClickListener(dateFilterButtonClickListener);
        binding.dailyRangeBtn.setOnClickListener(dateFilterButtonClickListener);
        binding.weeklyRangeBtn.setOnClickListener(dateFilterButtonClickListener);
        binding.monthlyRangeBtn.setOnClickListener(dateFilterButtonClickListener);
        binding.yearlyRangeBtn.setOnClickListener(dateFilterButtonClickListener);

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                String period = SharedPreferenceUtils.getInstance(getActivity()).getString(Constants.PREF_PERIOD, Constants.KEY_DAILY);
                switch (period) {
                    case Constants.KEY_DAILY:
                        binding.dailyRangeBtn.performClick();
                        binding.dailyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                        break;
                    case Constants.KEY_WEEKLY:
                        binding.weeklyRangeBtn.performClick();
                        binding.weeklyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                        break;
                    case Constants.KEY_MONTHLY:
                        binding.monthlyRangeBtn.performClick();
                        binding.monthlyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                        break;
                    case Constants.KEY_YEARLY:
                        binding.yearlyRangeBtn.performClick();
                        binding.yearlyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                        break;
                }
            }
        }, 500);

        return binding.getRoot();
    }

    private void resetDateRangeBtns(){
        binding.dailyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_unselected);
        binding.weeklyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_unselected);
        binding.monthlyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_unselected);
        binding.yearlyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_unselected);

    }

    private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = item -> {

        switch (item.getItemId()) {
            case R.id.navigation_home:
                binding.viewpager.setCurrentItem(0, true);
                actionBar.setTitle(tabTitles[0]);
                return true;
            case R.id.navigation_transaction:
                binding.viewpager.setCurrentItem(1, true);
                actionBar.setTitle(tabTitles[1]);
                return true;
            case R.id.navigation_accounts:
                binding.viewpager.setCurrentItem(2, true);
                actionBar.setTitle(tabTitles[2]);
                return true;
            case R.id.navigation_dashboard:
                binding.viewpager.setCurrentItem(3, true);
                actionBar.setTitle(tabTitles[3]);
                return true;

        }
        return false;
    };

    private ViewPager.OnPageChangeListener viewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

        }

        @Override
        public void onPageSelected(int position) {
            actionBar.setTitle(tabTitles[position]);

            if (dateContainerHeight == -1) {
                dateContainerHeight = binding.dateRangeContainer.getMeasuredHeight();
            }

            switch (position) {
                case 0: {
                    if (binding.dateRangeContainer.getMeasuredHeight() == 0) {
                        ValueAnimator anim = ValueAnimator.ofInt(0, dateContainerHeight);
                        anim.addUpdateListener(valueAnimator -> {
                            int val = (Integer) valueAnimator.getAnimatedValue();
                            ViewGroup.LayoutParams layoutParams = binding.dateRangeContainer.getLayoutParams();
                            layoutParams.height = val;
                            binding.dateRangeContainer.setLayoutParams(layoutParams);
                        });
                        anim.setDuration(200);
                        anim.start();

                        new Handler().postDelayed(() -> {
                            binding.dateRangeContainer.animate().alpha(1.0f).translationY(0).setDuration(200);
                        }, 100);
                    }

                    binding.navigation.setSelectedItemId(R.id.navigation_home);
                    break;
                }
                case 1: {
                    if (binding.dateRangeContainer.getMeasuredHeight() == 0) {
                        ValueAnimator anim = ValueAnimator.ofInt(0, dateContainerHeight);
                        anim.addUpdateListener(valueAnimator -> {
                            int val = (Integer) valueAnimator.getAnimatedValue();
                            ViewGroup.LayoutParams layoutParams = binding.dateRangeContainer.getLayoutParams();
                            layoutParams.height = val;
                            binding.dateRangeContainer.setLayoutParams(layoutParams);
                        });
                        anim.setDuration(200);
                        anim.start();

                        new Handler().postDelayed(() -> {
                            binding.dateRangeContainer.animate().alpha(1.0f).translationY(0).setDuration(200);
                        }, 100);
                    }

                    binding.navigation.setSelectedItemId(R.id.navigation_transaction);
                    break;
                }
                case 2: {
                    if (binding.dateRangeContainer.getMeasuredHeight() == dateContainerHeight) {
                        binding.dateRangeContainer.animate().alpha(0).translationY(-binding.dateRangeContainer.getHeight()).setDuration(200);
                        new Handler().postDelayed(() -> {
                            ValueAnimator anim = ValueAnimator.ofInt(dateContainerHeight, 0);
                            anim.addUpdateListener(valueAnimator -> {
                                int val = (Integer) valueAnimator.getAnimatedValue();
                                ViewGroup.LayoutParams layoutParams = binding.dateRangeContainer.getLayoutParams();
                                layoutParams.height = val;
                                binding.dateRangeContainer.setLayoutParams(layoutParams);
                            });
                            anim.setDuration(200);
                            anim.start();
                        }, 300);
                    }

                    binding.navigation.setSelectedItemId(R.id.navigation_accounts);
                    break;
                }

                case 3: {
                    if (binding.dateRangeContainer.getMeasuredHeight() == dateContainerHeight) {
                        binding.dateRangeContainer.animate().alpha(0).translationY(-binding.dateRangeContainer.getHeight()).setDuration(200);
                        new Handler().postDelayed(() -> {
                            ValueAnimator anim = ValueAnimator.ofInt(dateContainerHeight, 0);
                            anim.addUpdateListener(valueAnimator -> {
                                int val = (Integer) valueAnimator.getAnimatedValue();
                                ViewGroup.LayoutParams layoutParams = binding.dateRangeContainer.getLayoutParams();
                                layoutParams.height = val;
                                binding.dateRangeContainer.setLayoutParams(layoutParams);
                            });
                            anim.setDuration(200);
                            anim.start();
                        }, 300);
                    }

                    binding.navigation.setSelectedItemId(R.id.navigation_dashboard);
                    break;
                }
            }

        }

        @Override
        public void onPageScrollStateChanged(int state) {

        }
    };

    public void setActionBar(ActionBar supportActionBar) {
        this.actionBar = supportActionBar;
        actionBar.setTitle(getString(R.string.home));
    }

    @Override
    public void onDateChanged(int buttonId, String date, long startTime, long endTime) {
        if(buttonId != R.id.nextDateBtn && buttonId != R.id.previousDateBtn){
            resetDateRangeBtns();
        }

        switch (buttonId){
            case R.id.dailyRangeBtn:
                binding.dailyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                break;
            case R.id.weeklyRangeBtn:
                binding.weeklyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                break;
            case R.id.monthlyRangeBtn:
                binding.monthlyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                break;
            case R.id.yearlyRangeBtn:
                binding.yearlyRangeBtn.setBackgroundResource(R.drawable.bg_date_range_btn_selected);
                break;

        }

        binding.dateTv.setText(date);
        if (homeFragment != null && allExpenseFragment != null) {
            homeFragment.updateChartView(startTime, endTime);
            allExpenseFragment.updateFilterListWithDate(startTime, endTime, buttonId);
        }
    }

    public static class MainFragmentPagerAdapter extends FragmentPagerAdapter {

        public MainFragmentPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    homeFragment = new HomeFragment();
                    return homeFragment;
                case 1:
                    allExpenseFragment = new AllExpenseFragment();
                    return allExpenseFragment;
                case 2:
                    accountsFragment = new AccountsFragment();
                    return accountsFragment;
                case 3:
                    dashboardFragment = new DashboardFragment();
                    return dashboardFragment;
                default:
                    return null;
            }

        }

        @Override
        public int getCount() {
            return 4;
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return null;
        }
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/main/MainFragment.kt
package com.remotearthsolutions.expensetracker.fragments.main

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.content.res.Resources
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.ActionBar
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.activities.main.MainActivity
import com.remotearthsolutions.expensetracker.databinding.FragmentMainBinding
import com.remotearthsolutions.expensetracker.fragments.*
import com.remotearthsolutions.expensetracker.utils.Constants
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils
import com.remotearthsolutions.expensetracker.utils.findViewPagerFragmentByTag
import com.remotearthsolutions.expensetracker.views.PeriodButton

class MainFragment : BaseFragment(),
    DateFilterButtonClickListener.Callback {
    private var binding: FragmentMainBinding? = null
    private var pagerAdapter: MainFragmentPagerAdapter? = null
    private var actionBar: ActionBar? = null
    private var tabTitles: Array<String>? = null
    private var dateContainerHeight = -1
    private var selectedPeriodBtn: PeriodButton? = null
    private lateinit var mContext: Context
    private lateinit var mResources: Resources
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
        mResources = mContext.resources
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(
            inflater,
            R.layout.fragment_main,
            container,
            false
        )
        return binding?.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        registerBackButton(object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                (activity as MainActivity).onBackButtonPressed()
            }
        })

        tabTitles = arrayOf(
            getString(R.string.title_home),
            getString(R.string.title_transaction),
            getString(R.string.title_overview),
            getString(R.string.title_accounts)
        )

        pagerAdapter = MainFragmentPagerAdapter(childFragmentManager)

        binding!!.viewpager.offscreenPageLimit = 4
        binding!!.viewpager.adapter = pagerAdapter
        binding!!.viewpager.addOnPageChangeListener(viewPagerPageChangeListener)
        binding!!.navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)
        val dateFilterButtonClickListener = DateFilterButtonClickListener(this)
        binding!!.nextDateBtn.setOnClickListener(dateFilterButtonClickListener)
        binding!!.previousDateBtn.setOnClickListener(dateFilterButtonClickListener)
        binding!!.dailyRangeBtn.setOnClickListener(dateFilterButtonClickListener)
        binding!!.weeklyRangeBtn.setOnClickListener(dateFilterButtonClickListener)
        binding!!.monthlyRangeBtn.setOnClickListener(dateFilterButtonClickListener)
        binding!!.yearlyRangeBtn.setOnClickListener(dateFilterButtonClickListener)
        Handler().postDelayed({
            val period = SharedPreferenceUtils.getInstance(mContext)!!.getString(
                Constants.PREF_PERIOD,
                mResources.getString(R.string.daily)
            )
            val pos =
                listOf(*mResources.getStringArray(R.array.TimePeriod))
                    .indexOf(period)
            when (pos) {
                0 -> {
                    selectedPeriodBtn = binding!!.dailyRangeBtn
                }
                1 -> {
                    selectedPeriodBtn = binding!!.weeklyRangeBtn
                }
                2 -> {
                    selectedPeriodBtn = binding!!.monthlyRangeBtn
                }
                3 -> {
                    selectedPeriodBtn = binding!!.yearlyRangeBtn
                }
                else -> {
                    selectedPeriodBtn = binding!!.dailyRangeBtn
                    SharedPreferenceUtils.getInstance(mContext)!!.putString(
                        Constants.PREF_PERIOD,
                        resources.getString(R.string.daily)
                    )
                }
            }
            selectedPeriodBtn?.performClick()
        }, 500)
    }

    private fun resetDateRangeBtns() {
        binding!!.dailyRangeBtn.setIsSelected(false)
        binding!!.weeklyRangeBtn.setIsSelected(false)
        binding!!.monthlyRangeBtn.setIsSelected(false)
        binding!!.yearlyRangeBtn.setIsSelected(false)
    }

    private val mOnNavigationItemSelectedListener =
        BottomNavigationView.OnNavigationItemSelectedListener { item: MenuItem ->
            when (item.itemId) {
                R.id.navigation_home -> {
                    binding!!.viewpager.setCurrentItem(0, true)
                    actionBar?.title = tabTitles?.get(0)
                }
                R.id.navigation_transaction -> {
                    binding!!.viewpager.setCurrentItem(1, true)
                    actionBar?.title = tabTitles?.get(1)
                }
                R.id.navigation_overview -> {
                    binding!!.viewpager.setCurrentItem(2, true)
                    actionBar?.title = tabTitles?.get(2)
                }
                R.id.navigation_accounts -> {
                    binding!!.viewpager.setCurrentItem(3, true)
                    actionBar?.title = tabTitles?.get(3)
                }
            }
            true
        }

    private val viewPagerPageChangeListener: OnPageChangeListener = object : OnPageChangeListener {
        override fun onPageScrolled(
            position: Int,
            positionOffset: Float,
            positionOffsetPixels: Int
        ) {
        }

        override fun onPageSelected(position: Int) {
            actionBar?.title = tabTitles?.get(position)
            if (dateContainerHeight == -1) {
                dateContainerHeight = binding!!.dateRangeContainer.measuredHeight
            }
            when (position) {
                0 -> {
                    if (binding!!.dateRangeContainer.measuredHeight == 0) {
                        val anim = ValueAnimator.ofInt(0, dateContainerHeight)
                        anim.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Int
                            val layoutParams =
                                binding!!.dateRangeContainer.layoutParams
                            layoutParams.height = `val`
                            binding!!.dateRangeContainer.layoutParams = layoutParams
                        }
                        anim.duration = 200
                        anim.start()
                        anim.addListener(object : Animator.AnimatorListener {
                            override fun onAnimationStart(animation: Animator) {}
                            override fun onAnimationEnd(animation: Animator) {
                                val homeFragment =
                                    childFragmentManager.findViewPagerFragmentByTag<HomeFragment>(
                                        R.id.viewpager,
                                        0
                                    )
                                homeFragment?.refresh()
                            }

                            override fun onAnimationCancel(animation: Animator) {}
                            override fun onAnimationRepeat(animation: Animator) {}
                        })
                        Handler().postDelayed({
                            binding!!.dateRangeContainer.animate().alpha(1.0f).translationY(0f)
                                .duration = 200
                        }, 100)
                    }
                    binding!!.navigation.selectedItemId = R.id.navigation_home
                }
                1 -> {
                    if (binding!!.dateRangeContainer.measuredHeight == 0) {
                        val anim = ValueAnimator.ofInt(0, dateContainerHeight)
                        anim.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Int
                            val layoutParams =
                                binding!!.dateRangeContainer.layoutParams
                            layoutParams.height = `val`
                            binding!!.dateRangeContainer.layoutParams = layoutParams
                        }
                        anim.duration = 200
                        anim.start()
                        Handler().postDelayed({
                            binding!!.dateRangeContainer.animate().alpha(1.0f).translationY(0f)
                                .duration = 200
                        }, 100)
                    }
                    binding!!.navigation.selectedItemId = R.id.navigation_transaction
                }
                2 -> {
                    if (binding!!.dateRangeContainer.measuredHeight == 0) {
                        val anim = ValueAnimator.ofInt(0, dateContainerHeight)
                        anim.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Int
                            val layoutParams =
                                binding!!.dateRangeContainer.layoutParams
                            layoutParams.height = `val`
                            binding!!.dateRangeContainer.layoutParams = layoutParams
                        }
                        anim.duration = 200
                        anim.start()
                        Handler().postDelayed({
                            binding!!.dateRangeContainer.animate().alpha(1.0f).translationY(0f)
                                .duration = 200
                        }, 100)
                    }
                    binding!!.navigation.selectedItemId = R.id.navigation_overview
                }
                3 -> {
                    if (binding!!.dateRangeContainer.measuredHeight == dateContainerHeight) {
                        binding!!.dateRangeContainer.animate().alpha(0f)
                            .translationY(-binding!!.dateRangeContainer.height.toFloat()).duration =
                            200
                        Handler().postDelayed({
                            val anim = ValueAnimator.ofInt(dateContainerHeight, 0)
                            anim.addUpdateListener { valueAnimator: ValueAnimator ->
                                val `val` = valueAnimator.animatedValue as Int
                                val layoutParams =
                                    binding!!.dateRangeContainer.layoutParams
                                layoutParams.height = `val`
                                binding!!.dateRangeContainer.layoutParams = layoutParams
                            }
                            anim.duration = 200
                            anim.start()
                        }, 300)
                    }
                    binding!!.navigation.selectedItemId = R.id.navigation_accounts
                }
            }
        }

        override fun onPageScrollStateChanged(state: Int) {}
    }

    override fun onDateChanged(
        btnId: Int,
        date: String?,
        startTime: Long,
        endTime: Long
    ) {
        if (btnId != R.id.nextDateBtn && btnId != R.id.previousDateBtn) {
            resetDateRangeBtns()
        }
        when (btnId) {
            R.id.dailyRangeBtn -> {
                selectedPeriodBtn = binding!!.dailyRangeBtn
            }
            R.id.weeklyRangeBtn -> {
                selectedPeriodBtn = binding!!.weeklyRangeBtn
            }
            R.id.monthlyRangeBtn -> {
                selectedPeriodBtn = binding!!.monthlyRangeBtn
            }
            R.id.yearlyRangeBtn -> {
                selectedPeriodBtn = binding!!.yearlyRangeBtn
            }
        }
        selectedPeriodBtn?.setIsSelected(true)

        binding!!.dateTv.text = date
        val homeFragment =
            childFragmentManager.findViewPagerFragmentByTag<HomeFragment>(R.id.viewpager, 0)
        val allExpenseFragment =
            childFragmentManager.findViewPagerFragmentByTag<AllExpenseFragment>(R.id.viewpager, 1)
        if (homeFragment != null && allExpenseFragment != null) {
            homeFragment.updateChartView(startTime, endTime)
            allExpenseFragment.updateFilterListWithDate(
                startTime,
                endTime,
                btnId
            )
        }
        (mContext as MainActivity).updateSummary(startTime, endTime)
    }

    fun refreshChart() {
        if (selectedPeriodBtn != null) {
            selectedPeriodBtn!!.performClick()
        }
    }

    class MainFragmentPagerAdapter(fm: FragmentManager?) :
        FragmentPagerAdapter(fm!!, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
        override fun getItem(position: Int): Fragment {
            return when (position) {
                0 -> {
                    val homeFragment = HomeFragment()
                    homeFragment
                }
                1 -> {
                    val allExpenseFragment = AllExpenseFragment()
                    allExpenseFragment
                }
                2 -> {
                    val overViewFragment = OverViewFragment()
                    overViewFragment
                }
                3 -> {
                    val accountsFragment = AccountsFragment()
                    accountsFragment
                }
                else -> {
                    Fragment()
                }
            }
        }

        override fun getCount(): Int {
            return 4
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return null
        }
    }

//    private fun <T : Fragment> findViewPagerFragmentByTag(
//        fragmentManager: FragmentManager,
//        fragmentPosition: Int
//    ): T? {
//        val fragmentTag = "android:switcher:${R.id.viewpager}:${fragmentPosition}"
//        return fragmentManager.findFragmentByTag(fragmentTag) as T?
//    }

}